app/models/data_set.rb

```diff
@@ -101,31 +97,52 @@ def processing_complete?
   end
 
   def data_file=(file)
-    self.csv_data = read_as_utf8(file)
-
-    @need_csv_processing = true
+    if file.nil?
+      @csv_data = nil
+      @need_csv_processing = false
+    else
+      @csv_data = CsvData.new(data_file: file)
+      @need_csv_processing = true
+    end
   end
 
   def schedule_csv_processing
     if @need_csv_processing
+      @csv_data.save!
       ProcessCsvDataWorker.perform_async(service.id.to_s, self.version)
-      @need_csv_processing = false
+      @need_csv_processing = nil
```

Comment:
Oops.  Left over from a previous attempt where `@need_csv_processing` was the `CsvData` object to be processed, rather than a flag (similar to how it is now in `Service`).  I'll change this back, thanks!